x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality_panel/data_quality_details/indices_details/pattern/index.tsx

```diff
@@ -242,92 +255,87 @@ const PatternComponent: React.FC<Props> = ({
     httpFetch,
     formatBytes,
     formatNumber,
+    flyoutChartSelectedActionAbortControllerRef,
   ]);
 
-  useEffect(() => {
-    const flyoutIndexExpandActionAbortController =
-      flyoutIndexExpandActionAbortControllerRef.current;
-    const tableRowIndexCheckNowActionAbortController =
-      tableRowIndexCheckNowActionAbortControllerRef.current;
-    const flyoutIndexChartSelectedActionAbortController =
-      flyoutIndexChartSelectedActionAbortControllerRef.current;
-    return () => {
-      flyoutIndexExpandActionAbortController.abort();
-      tableRowIndexCheckNowActionAbortController.abort();
-      flyoutIndexChartSelectedActionAbortController.abort();
-    };
-  }, []);
-
   return (
     <div data-test-subj={`${pattern}PatternPanel`}>
-      <PatternAccordion
-        id={patternComponentAccordionId}
-        initialIsOpen={true}
-        buttonElement="div"
-        buttonContent={
-          <PatternSummary
-            incompatible={getTotalPatternIncompatible(patternRollup?.results)}
-            indices={indexNames?.length}
-            indicesChecked={getTotalPatternIndicesChecked(patternRollup)}
-            ilmExplainPhaseCounts={ilmExplainPhaseCounts}
-            pattern={pattern}
-            patternDocsCount={patternRollup?.docsCount ?? 0}
-            patternSizeInBytes={patternRollup?.sizeInBytes}
-          />
-        }
+      <HistoricalResultsContext.Provider
+        value={{
+          fetchHistoricalResults,
+          historicalResultsState,
```

Comment:
It is not a very good practice to use react context for state management. I understand the problem it solves here and I'm not asking for a refactor now, but we usually limit its usage to properties that never change.

Reply:
Interesting perspective @semd

But here are my thoughts if you are interested:

1. DQ Dashboard didn't have official state management solution and for these use cases it would be imho unnecessary and arguably an overkill to introduce redux. And frankly I don't think I had time :D
2. Now about "best practice". It depends. As outlined in https://react.dev/learn/passing-data-deeply-with-context#use-cases-for-context explicitly 
![image](https://github.com/user-attachments/assets/71d1cfa3-669d-4a97-888f-b7618c043286)
context was developed as state management solution and if you noticed I used it exactly with reducers, but of course everyone has their own experience with regards to best practices and I totally understand why you would say that. But I would argue that context is absolutely viable in this architecture and shouldn't be frowned upon. But I would be glad to discuss this in a bigger setup. Ideally when it gets more complex I would probably lean more towards redux or a little more lightweight solutions that have proper memoization builtin and selectors. But for this use case it was pretty good and it has tremendously simplified both source code and test code. Before this prod drilling was absolutely out of hand.

Let me know wdyt.

Reply:
@kapral18 
> DQ Dashboard didn't have official state management solution and for these use cases it would be imho unnecessary and arguably an overkill to introduce redux. And frankly I don't think I had time :D

Agree, no need to introduce redux here.

> Now about "best practice". It depends

Yes, it depends. context is not a state management tool but a tool to avoid prop drilling. It can be used for state management, yes, but it has to be used with caution. As you know, the problem with it is that it re-renders **all** the components that consumes it, regardless of the props they retrieve.

In this specific implementation, all the components that only use `fetchHistoricalResults` are also re-rendered when `historicalResultsState` changes. This is not a big problem, I agree, but in other cases, depending on how it is used, it may cause performance issues.

<details>
  <summary>Here's a diagram that I found very useful some time ago:</summary>
  
![context_usage_diagram](https://github.com/user-attachments/assets/3a1f25c6-7cd0-4a0e-b5d4-6f83f36932bb)

</details>

The main issue with context+state is that it's not an easy task to determine if is a good or bad choice, the reviewer needs to go deep into the components workflow to understand the amount of rerenders it will cause, and even then it can still be a matter of discussion. Also, it may not be a problem when it is introduced, but may become a problem when the application grows. That's why we have conventions, to avoid spending too much time discussing the same topic everytime it appears, but have an opinionated way to go.

Honestly I haven't found anything writer about this convention, I have only seen some comments in some PRs, and I also haven't found any context with state management in Kibana, they are all used basically to store values that do not (often) change.

About this specific implementation:

Since I had to check it, I noticed that there are only 2 components that consumes `historicalResultsState` (and a hook that is only used by one of these components). I don't think passing the state values as props is a big deal here, we already do this with other properties such as `indexName`.

Reply:
Well again you are basing your argument entirely on rerenders. If I wrap (and I am not against it) the value in a useMemo) the problem is gone. It's not a gigantic table with a million live rerendered nodes that we should care about optimizing every single ms from. We are probably arguing about a couple of ms unoptimized or no  delay at all with useMemo.

Also I am not saying context + reducer that I use is better than redux or zustand or react-query obviously it isn't, it's  a barebones state-management which perfectly fits for the initially considered use case.

Now you well noticed that it's only used in 2 places, which wasn't the case during the implementation :) so you could say we can cut it out. But if we are talking about future and convenience, the thing is you can prop drill but that's how you eventually end up with what I started with 10 levels of props being passed down. I doubt they started like that. Probably every next person was adding +1 layer down to not refactor the entire thing.

And the thing that makes me think that keeping context for passing this down useful is that I see historical results and fetchHistoricalResults as  something that will be used on all or most levels of the historical results domain. So to me it's a domain specific functionality. There is other functionality which is very narrow and specific to a given component that I wouldn't pass into context.

But I want to ask you before you noticed that there were only 2 usages. What was on your mind as an alternative to the context for historical results? prop drilling or something else?

Reply:
Wrapped in useMemo with latest commit.

Reply:
Yes I completely understand your argument, and having a correct component memoization is key, I agree.

Unfortunately, historically here, we've run into issues where devs making downstream changes introduce code that seemingly preforms well in their PR and in testing, but when run at scale runs into problems specifically because of lack of rendering stability upstream. Because of that, we adopt defensive strategies to account for these scenarios.

I also noticed that the same approach has been followed with `ResultsRollupContext` and `IndicesCheckContext`. I am not saying this is wrong, having separate state context is better than having only a big one, I acknowledge that. But that does not solve the underlaying problem mentioned above.

And I am not asking for a change in this PR. There's no need to refactor these contexts right now, FF is around the corner, and it won't represent any noticeable performance improvement, since the components implementation is robust. The aim was only to rise awareness, discuss this topic, and, if we agree, change it later on as tech debt.

> Before noticing there were only two usages, what alternative to using context for historical results did you have in mind? Prop drilling or something else?

I won't choose a specific solution without having a clear idea of the component hierarchy and the state shape. Personally, what I do is consider passing separate props first (not only one big state object prop, and trying to avoid creating it too high in the component hierarchy), if this strategy becomes too cumbersome, then I prefer creating a Redux store slice. Other people prefer to use ReactQuery as an intermediate solution, that's also a good option.

Let's resolve this comment, and discuss it synchronously, if you want. Does it sound good?